### PR TITLE
Refactor Modal Visible parameter and state management

### LINKED ISSUE #1359: Modal - Show event
Hello,

On you [Model documentation](https://blazorise.com/docs/components/modal/#closing) you have a closing event, you could also have a `Showing` and/or `Shown`.  
This is a requirement for what I want to use the models for, currently in the app I am developing I use `RenderFragment` for my dialogs, so every time a dialog shows then the component does it's `Init`, inside the init method I can then setup the form and the object the form is editing.

I am looking to replace my `RenderFagments` with models in the hop to improve performance.

What do you think?

### LINKED ISSUE #2810: Modal Closing is an Action instead of an EventCallback
Model Closing event is of type `Action<ModalClosingEventArgs>` so when I want to do something async in this OnClosing it will have to be something like this `async void OnClosing(ModalClosingEventArgs args) { ....  }` which is undesirable. The type of this event should be an EventCallback so async operations are supported.

### PR DESCRIPTION
**Pull request by stsrki**
closes #2810 Modal Closing is an Action instead of an EventCallback
closes #1359 Modal - Show event

I have refactored the `Visible` parameter and `Closing` event. Instead of doing all the logic in the `Visible` setter, I have moved it to the `SetParametersAsync` lifecycle.

- The benefit of this is that the `Closing` event can now be properly awaited.
- The disadvantage is that `Visible` is used only as a hint for us that the parameter was received by the Blazot lifecycle. It should never be used directly as that would leave it in a bad state, eg. `modal.Visible = true`.

This will also come in handy for #1359

### CODE REVIEW

File: Source/Blazorise.Bulma/wwwroot/blazorise.bulma.js [hunk 1]
```diff
@@ -24,7 +24,11 @@ window.blazoriseBulma = {
     modal: {
         open: (element, scrollToTop) => {
             if (scrollToTop) {
-                element.querySelector('.modal-card-body').scrollTop = 0;
+                const modalBody = element.querySelector('.modal-card-body');
```
Comment (David-Moreira) [inline Source/Blazorise.Bulma/wwwroot/blazorise.bulma.js:27]: Forgot material.js?

File: Source/Blazorise/Components/Modal/Modal.razor.cs [hunk 2]
```diff
@@ -61,6 +61,32 @@ public partial class Modal : BaseComponent, ICloseActivator
...
+                    HandleVisibilityStyles( true );
+                    RaiseEvents( true );
+                }
+                else if ( visibleResult == false && await IsSafeToClose() )
+                {
+                    state = state with { Visible = false };
+
+                    HandleVisibilityStyles( false );
+                    RaiseEvents( false );
+                }
```
Comment (David-Moreira) [inline Source/Blazorise/Components/Modal/Modal.razor.cs]: If these are always done in 'bulk' maybe you can extract these into a method with a bool argument?

**Review by David-Moreira — CHANGES_REQUESTED** [review]
Tested on bootstrap, looking good.

Update Doc Attributes, `Closing` is now wrong? Also missing alot of attributes.
image

Two async calls, setup Async?
image

Could `Opening` and `Closing` be EventCallbacks? Or you think it's not worth it?
Also if there's a `Closed` EventCallback, maybe we should introduce an `Opened` EventCallback?
image

File: Source/Blazorise.Bulma/wwwroot/blazorise.bulma.js (hunk 1 above)
Comment (stsrki) [reply Source/Blazorise.Bulma/wwwroot/blazorise.bulma.js:27]: Done

**Comment by njannink** [comment]
Since setting the Visible paramater isn't really a good idea, because it will leave the component in a bad state. Wouldnt it be better to make the setter private or protected?

**Comment by stsrki** [comment]
> Since setting the Visible parameter isn't really a good idea, because it will leave the component in a bad state. Wouldnt it be better to make the setter private or protected?

Every parameter must be made public and that is Blazor constraint. Otherwise, it will not compile. With the changes in this PR the use of the `Visible` parameter is made in the right way. Meaning it can only be defined declaratively. And if you need to define it imperatively, you still have the option to use `Show()` and `Hide()`.

**Pull request merged**